docs(proposal): add KubeVirt GPU NUMA proposal

jimkingstone opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
To optimize computing performance in KubeVirt GPU virtual machines.

**Issue**:
https://github.com/kubevirt/kubevirt/issues/13926

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Support associating NUMA nodes with GPU devices
```

jimkingstone commented:
Migrated from https://github.com/kubevirt/community/pull/394 @lyarwood

iholder101 commented:
Hey @jimkingstone! Thanks for this proposal.

Please read the instructions on how to propose VEPs here: https://github.com/kubevirt/enhancements/blob/main/README.md. For example, please create a tracking issue.

In addition, please use the github template format for this PR instead of copy-pasting the kubevirt/community template.

Thanks.

iholder101 commented:
/hold

jimkingstone commented:
> Hey @jimkingstone! Thanks for this proposal.
> 
> Please read the instructions on how to propose VEPs here: https://github.com/kubevirt/enhancements/blob/main/README.md. For example, please create a tracking issue.
> 
> In addition, please use the github template format for this PR instead of copy-pasting the kubevirt/community template.
> 
> Thanks.

OK, the issue has been created, and this PR template has been updated.

alaypatel07 commented:
/cc 
/assign

I am interested in reviewing this proposal as this could have overlapping solutions with DRA gpu devices

cc @rthallisey

vladikr commented:
Thank you @jimkingstone for looking into this issue.

First, I'm not 100% sure if we need a formal design proposal to address this, mainly because there is no change in KubeVirt API.

That's, indeed, an issue we have been trying to resolve for a while. 
In the past, I tried to expose the NUMA-related (collection of the vCPUs associated with a specific numa node - in the case without guest numa mapping) information using device metadata: https://github.com/kubevirt/kubevirt/pull/6946

My primary concern here is that so far KubeVirt has delegated the creation of controllers to libvirt. This can get very complex, very quickly especially when there are multiple different devices involved.  How confident are we that we want to open this door?

/cc @alicefr @andreabolognani what do you think?

jimkingstone commented:
> Thank you @jimkingstone for looking into this issue.
> 
> First, I'm not 100% sure if we need a formal design proposal to address this, mainly because there is no change in KubeVirt API.
> 
> That's, indeed, an issue we have been trying to resolve for a while. In the past, I tried to expose the NUMA-related (collection of the vCPUs associated with a specific numa node - in the case without guest numa mapping) information using device metadata: [kubevirt/kubevirt#6946](https://github.com/kubevirt/kubevirt/pull/6946)
> 
> My primary concern here is that so far KubeVirt has delegated the creation of controllers to libvirt. This can get very complex, very quickly especially when there are multiple different devices involved. How confident are we that we want to open this door?
> 
> /cc @alicefr @andreabolognani what do you think?

Thanks to @vladikr for the suggestion.

As for whether we need a formal proposal, I’ve also submitted the related code changes(https://github.com/kubevirt/kubevirt/pull/14406), which can be reviewed alongside the proposal.

Regarding whether we should directly use the topology file — as mentioned in the proposal, _“Compared to directly providing a topology file, this approach more accurately reflects the physical GPU NUMA relationship and is more friendly to NUMA topology awareness because users do not need to manually obtain and depend on a topology file.”_

This change does introduce multiple different devices. I’ve done some initial testing on an L40s machine, and the results (as shown in the screenshot https://github.com/kubevirt/kubevirt/issues/13926) are as expected. However, more testing on different machine models may be needed for further validation.

lyarwood commented on veps/sig-compute/gpu-numa.md:
````diff
@@ -0,0 +1,278 @@
+
+```bash
+cat /sys/bus/pci/devices/0000\:60\:00.0/numa_node
+0
````
Apologies if this is a stupid question but is the device guaranteed to be present in the same pNUMA node as the pCPUs we've been scheduled to? I'm assuming so otherwise none of this would work but I have no idea if that's done by the device plugin, cpu, memory, topology manager etc.

lyarwood commented on veps/sig-compute/gpu-numa.md:
````diff
@@ -0,0 +1,278 @@
+        memory:              # memory NUMA settings
+          hugepages:         # (KubeVirt already supported)
+            pageSize: "2Mi"  #
+```
````
You need to request a GPU in the VM still right?

jimkingstone replied:
Yes, sorry for the oversight. The GPU settings in the VM have been added.

lyarwood commented on veps/sig-compute/gpu-numa.md:
```diff
@@ -0,0 +1,278 @@
+  configuration:
+    developerConfiguration:
+      featureGates:
+      - GPUDeviceNUMA # the feature gate to control the enabling and disabling of this functionality
```
The feature gate is a temporary way of enabling and disabling this functionality.

If we want cluster admins to be able to enable/disable or otherwise configure the feature then an additional set of configurables should be introduced somewhere.

jimkingstone replied:
Can we expose a VM annotation to the creators to control the feature enable/disable?

jimkingstone commented on veps/sig-compute/gpu-numa.md:
```diff
@@ -0,0 +1,278 @@
+            <cell id="1" cpus="40,41,42,43,44,45,46,47" memory="32212254720" unit="b"></cell>
+        </numa>
+    </cpu>
+...
```
Thx, it's been added.

andreabolognani commented:
@jimkingstone @vladikr I've given this proposal and the corresponding PR a cursory look.

IIUC the main idea is that, when the existing `cpu.numa.guestMappingPassthrough` feature is enabled, the guest NUMA topology is going to be configured so that it matches the host NUMA topology.

When the feature you're introducing is enabled then, KubeVirt will go the extra mile and figure out the host NUMA node for each GPU, look up the corresponding guest NUMA node based on that information, and ensure that a PCIe Expander Bus assigned to the correct guest NUMA node is used to attach the GPU.

Seems fairly reasonable overall, and certainly desirable when it comes to performance. There are a few things that I'm concerned/wondering about though.

Have you made sure that these PCIe controllers, for which you're allocating indexes explicitly, are always added to the domain XML before any other PCIe controller? Failing to do so might trip up libvirt's PCI address allocation logic.

AFAIK PCIe Expander Buses, just like most other PCIe controllers, can't be hot(un)plugged. Are assigned devices generally expected to be hot(un)pluggable in KubeVirt? If so, this might impose a new limitation.

Why is this restricted to GPUs? It seems to me that it should apply to all assigned devices, including e.g. network adapters.

jimkingstone commented:
> @jimkingstone @vladikr I've given this proposal and the corresponding PR a cursory look.
> 
> IIUC the main idea is that, when the existing `cpu.numa.guestMappingPassthrough` feature is enabled, the guest NUMA topology is going to be configured so that it matches the host NUMA topology.
> 
> When the feature you're introducing is enabled then, KubeVirt will go the extra mile and figure out the host NUMA node for each GPU, look up the corresponding guest NUMA node based on that information, and ensure that a PCIe Expander Bus assigned to the correct guest NUMA node is used to attach the GPU.
> 
> Seems fairly reasonable overall, and certainly desirable when it comes to performance. There are a few things that I'm concerned/wondering about though.
> 
> Have you made sure that these PCIe controllers, for which you're allocating indexes explicitly, are always added to the domain XML before any other PCIe controller? Failing to do so might trip up libvirt's PCI address allocation logic.
> 
> AFAIK PCIe Expander Buses, just like most other PCIe controllers, can't be hot(un)plugged. Are assigned devices generally expected to be hot(un)pluggable in KubeVirt? If so, this might impose a new limitation.
> 
> Why is this restricted to GPUs? It seems to me that it should apply to all assigned devices, including e.g. network adapters.

Thanks to @andreabolognani  for the questions. Apologies for the delay in responding.

1. Yes, these PCIe controllers will be explicitly added, with index and bus numbers allocated, and we will ensure the order of the indexes. The remaining PCIe controllers will be automatically allocated by libvirt.

2. Hot-plugging is not supported at the moment. I have added this limitation to the "Non Goals" section of the proposal.

3. Currently, we only apply this to GPU devices, i.e., based on `vm.spec.template.spec.domain.devices.gpus`, to determine whether a device is associated with a NUMA node. I have added this limitation to both the "Non Goals" and "API Examples" sections of the proposal.

jean-edouard commented on veps/sig-compute/gpu-numa.md:
```diff
@@ -0,0 +1,292 @@
+
+# Design
+
+We propose using emulating a `pcie-expander-bus (pxb-pcie)` in the VM to
```
```suggestion
We propose emulating a `pcie-expander-bus (pxb-pcie)` in the VM to
```

jean-edouard commented on veps/sig-compute/gpu-numa.md:
```diff
@@ -0,0 +1,292 @@
+# Design
+
+We propose using emulating a `pcie-expander-bus (pxb-pcie)` in the VM to
+configure NUMA node and expose a `pcie-root-port` for PCIe device (GPU devices)
```
```suggestion
configure NUMA and expose a `pcie-root-port` for PCIe devices (GPU devices)
```

jean-edouard commented on veps/sig-compute/gpu-numa.md:
```diff
@@ -0,0 +1,292 @@
+We propose using emulating a `pcie-expander-bus (pxb-pcie)` in the VM to
+configure NUMA node and expose a `pcie-root-port` for PCIe device (GPU devices)
+to plug into, according to the [QEMU device placement strategy](
+https://github.com/qemu/qemu/blob/master/docs/pcie.txt#L37-L74):
```
```suggestion
https://github.com/qemu/qemu/blob/master/docs/pcie.txt#L37-L74).
```

jean-edouard commented on veps/sig-compute/gpu-numa.md:
```diff
@@ -0,0 +1,292 @@
+https://github.com/qemu/qemu/blob/master/docs/pcie.txt#L37-L74):
+
+PCIe endpoint devices are not themselves associated with NUMA nodes, rather the
+bus they are connected to has affinity. The root complex(pcie.0) is not
```
```suggestion
bus they are connected to has a NUMA affinity. The root complex(pcie.0) is not
```

jean-edouard commented on veps/sig-compute/gpu-numa.md:
```diff
@@ -0,0 +1,292 @@
+be added and associated with a NUMA node.
+
+It is not possible to plug PCIe endpoint devices directly into the
+`pcie-expander-bus (pxb-pcie)`, so it is necessary to add `pcie-root-port` into
```
```suggestion
`pcie-expander-bus (pxb-pcie)`, so it is necessary to add a `pcie-root-port` into
```

jean-edouard commented on veps/sig-compute/gpu-numa.md:
```diff
@@ -0,0 +1,292 @@
+
+## Update/Rollback Compatibility
+
+The feature gate `GPUDeviceNUMA` can control the enabling and disabling of
```
Please ignore the feature gate in this section. Feature-gated code is disabled by default, so it obviously doesn't affect updates. Please assume that the feature gate is enabled as part of the update. Can you think of any potential issue then?
